### PR TITLE
Improve performance of logits MatMul for first decoding run in BeamSearch

### PR DESCRIPTION
**Pull request by hariharans29**
### Description
In BeamSearch, for the first decoding run with the user-provided input sequence, the final logits MatMul projects the encoding of the entire batched sequence onto the vocab space (i.e.) the MatMul's input shapes are [B, S, H] * [H, vocab_size]. This is sub-optimal as we only need to project the encoding of the last token of the sequence in each batch (i.e.) the logits MatMul's inputs should be [B, 1, H] * [H, vocab_size] where the `1` will be the last token of that batch. Today, we do the costlier MatMul (using [B,S,H]) and then slice out the logits corresponding to the last token per-batch. 

To reduce the cost of the MatMul in the first decoding run, a new custom version of MatMul is proposed called `LastTokenMatMul` (the spec of which is defined to strictly adhere to something which we will see in decoding subgraphs) which will slice out the last token per-batch and then do the MatMul. This result will be consumed seamlessly in the `BeamSearch` kernel (no change required there). If the provided batch has only one token in it, the slicing is skipped and the operation becomes regular MatMul.

TODO:  Run some perf evaluations

### Motivation and Context
Improve performance of BeamSearch

### CODE REVIEW

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 1]
```diff
@@ -922,13 +965,29 @@ def convert_generation_model(args: argparse.Namespace, generation_type: Generati
...
+    # onto the vocabulary space. We hence replace the MatMul operator with a customized
+    # operator that does just that.
+    # NOTE: We currently only support this for fp16/fp32 GPT2 BeamSearch.
+    # This can be expanded to other models/decoding strategies/data types later
+    last_token_logits_matmul_enabled = False
+    if args.enable_last_token_logits_matmul and is_gpt2 and generation_type == GenerationType.BEAMSEARCH:
+        logger.info(
+            f"Replace logits MatMul with LastTokenMatMul on {args.decoder_onnx}. " "The file will be overwritten."
+        )
+        last_token_logits_matmul_enabled = replace_logits_matmul(args.decoder_onnx, args.use_external_data_format)
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_generation.py:981]: For GPT-2, the first decoding is done in decoder_onnx. For encoder-decoder model like T5, first decoding is done in encoder_decoder_init subgraph.

**Review by tianleiwu — COMMENTED** [review]
It's good idea for optimization.
Shall we just add a slice before last MatMul instead of adding a new operator?

**Comment by hariharans29** [comment]
> It's good idea for optimization. Shall we just add a slice before last MatMul instead of adding a new operator?

Thanks for the comment. That was my initial idea too - just adding a `Slice`. But `Slice` will need to populate the output buffer (i.e.) if the sequence length is 1 (which if my understanding is right, is always the case from the second iteration onwards), it will need to make a copy of the buffer. In this approach, we get to skip the slicing if seq_length == 1 (saves a copy) - @yufenglee was the one who pointed this out.

File: onnxruntime/python/tools/transformers/convert_generation.py (hunk 1 above)
Comment (hariharans29) [reply onnxruntime/python/tools/transformers/convert_generation.py:981]: For now, this method is only called when the model is GPT2 (and so was the previous optimization wrt to vocab padding). Would you like to expand this for T5 or prefer that we do it in a separate change later ?

**Review by yufenglee — APPROVED** [review]
:shipit:

**Comment by hariharans29** [comment]
> Is it possible to improve Slice to not copy (just returns the input buffer when output is exactly same as input?) in this case.


 Is it possible to improve Slice to not copy (just returns the input buffer when output is exactly same as input?) in this case.`

There are a couple of hints to the allocation planner that can trigger re-use of the input buffer:
1) Alias - This unconditionally re-uses the input buffer (used for ops like Reshape), we can't apply this to Slice as we can't statically determine if the output is going to tbe same as the input
2) MayInPlace - This uses the input as the output as long as the shapes of the input and output are the same and the use count of the input == 1 (used mainly for activation ops). It needs some more thought but we *may* be able to try this...

You can think of this approach as Slice getting inserted and then subsequently fused with the downstream MatMul, only instead of writing the fusion, we update the graph in place....

**Comment by yufenglee** [comment]
We can close this one.
